src/Microsoft.AspNetCore.Routing/Matching/RefEmitTrieJumpTable.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+#if NET462 || NETCOREAPP2_0
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace Microsoft.AspNetCore.Routing.Matching
+{
+    internal class RefEmitTrieJumpTable : JumpTable
+    {
+        private readonly int _defaultDestination;
+        private readonly int _exitDestination;
+        private readonly (string text, int destination)[] _entries;
+
+        private readonly Func<string, int, int, int> _delegate;
+
+        public RefEmitTrieJumpTable(
+            int defaultDestination,
+            int exitDestination,
+            (string text, int destination)[] entries)
+        {
+            _defaultDestination = defaultDestination;
+            _exitDestination = exitDestination;
+            _entries = entries;
+
+            _delegate = GenerateDelegate();
```

Comment:
Need to measure what the startup cost of this would be for a large site. Might be an idea to generate the IL lazy on first call to `GetDestination`.

Reply:
Yeah, definitely. We'll offload this to task and use it once its available.